File: rust/bench-common/src/serial.rs
//! Shared harness for the `serialization` focus area: env config, a counting
//! global allocator (deterministic decode-time memory measurement), and the
//! journal write/replay timed loop that emits the four result-contract metrics.
//!
//! stdout stays result-only; this module prints nothing but the emit lines.

#[cfg(test)]
#[global_allocator]
static TEST_ALLOC: CountingAllocator = CountingAllocator;

use std::alloc::{GlobalAlloc, Layout, System};
use std::hint::black_box;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::time::Instant;

use crate::{result, stats};

const FOCUS: &str = "serialization";

/// Env-configurable knobs. `SER_ENTRIES` tunes the entry count; `SER_CMD_BYTES`
/// sizes the `cmd_text` string on each entry (default 12). The record is
/// field-heavy, dominated by the typed scalar fields rather than cmd_text.
pub struct SerialConfig {
    pub warmup: usize,
    pub iterations: usize,
    pub entries: usize,
    pub cmd_bytes: usize,
}

impl SerialConfig {
    pub fn from_env() -> Result<SerialConfig, String> {
        Ok(SerialConfig {
            warmup: parse_env("SER_WARMUP", 1_000)?,
            iterations: parse_env("SER_ITERS", 100_000)?,
            entries: parse_env("SER_ENTRIES", 4)?,
            cmd_bytes: parse_env("SER_CMD_BYTES", 12)?,
        })
    }
}

fn parse_env(key: &str, default: usize) -> Result<usize, String> {
    match std::env::var(key) {
        Ok(v) => v
            .parse::<usize>()
            .map_err(|_| format!("{key}: expected a positive integer, got {v:?}")),
        Err(_) => Ok(default),
    }
}

// ---- counting global allocator ------------------------------------------------

static ALLOCATED: AtomicUsize = AtomicUsize::new(0);

/// Wraps the system allocator and sums the byte size of every allocation
/// request. Deterministic and repeatable — the memory signal the focus area
/// exists to compare. Install in each bench binary with `#[global_allocator]`.
pub struct CountingAllocator;

// SAFETY: every method delegates to `System`, which is itself a valid
// `GlobalAlloc` implementation; this wrapper only adds a non-mutating
// bookkeeping side effect (an atomic counter update) around each call and
// does not alter the pointers, layouts, or safety contract passed through.
unsafe impl GlobalAlloc for CountingAllocator {
    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
        ALLOCATED.fetch_add(layout.size(), Ordering::Relaxed);
        unsafe { System.alloc(layout) }
    }
    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
        unsafe { System.dealloc(ptr, layout) }
    }
    unsafe fn realloc(&self, ptr: *mut u8, layout: Layout, new_size: usize) -> *mut u8 {
        if new_size > layout.size() {
            ALLOCATED.fetch_add(new_size - layout.size(), Ordering::Relaxed);
        }
        unsafe { System.realloc(ptr, layout, new_size) }
    }
}

pub fn allocated_bytes() -> usize {
    ALLOCATED.load(Ordering::Relaxed)
}

pub fn reset_allocated() {
    ALLOCATED.store(0, Ordering::Relaxed);
}

// ---- journal write/replay harness --------------------------------------------

/// Encode-time then replay-time driver. `build(index) -> R` produces one logical
/// record deterministically; `encode(record, &mut scratch) -> len` serializes
/// into a caller buffer; `decode(bytes) -> checksum` decodes and **fully
/// materializes** (folds every field) so lazy codecs pay for the reads.
///
/// Generic over the record type `R` so this module (and `bench-common` as a
/// whole) stays serde-free and focus-neutral — the codec cell supplies the
/// concrete builder.
///
/// Emits, for both `encode` and `decode`: `<op>_p50`, `<op>_p99` (integer ns)
/// and `<op>_mean` (fractional ns) — mirroring the `network-rtt`
/// `rtt_{p50,p99,mean}` convention — plus `encoded_bytes` (one record) and
/// `decode_alloc_bytes` (heap bytes allocated per decode, via the counting
/// allocator). All samples/scratch are preallocated so the allocator counter
/// reflects codec allocation only.
pub fn run_journal<R, B, E, D>(
    experiment: &str,
    cfg: &SerialConfig,
    build: B,
    mut encode: E,
    mut decode: D,
) where
    B: Fn(u64) -> R,
    E: FnMut(&R, &mut [u8]) -> usize,
    D: FnMut(&[u8]) -> u64,
{
    let n = cfg.iterations;
    // Pre-build the records (untimed); building is deterministic from index.
    let records: Vec<R> = (0..(cfg.warmup + n) as u64).map(&build).collect();

    let mut scratch = vec![0u8; 64 * 1024];
    let mut encode_ns: Vec<u64> = Vec::with_capacity(n);
    let mut record_len = 0usize;

    // Warmup encode.
    for r in records.iter().take(cfg.warmup) {
        record_len = encode(r, &mut scratch);
    }
    // Timed encode.
    for r in records.iter().skip(cfg.warmup) {
        let t0 = Instant::now();
        let len = encode(r, &mut scratch);
        let dt = t0.elapsed().as_nanos() as u64;
        black_box(&scratch[..len]);
        record_len = len;
        encode_ns.push(dt);
    }

    // Build the contiguous in-memory journal from the timed records.
    let mut journal = Vec::with_capacity(record_len * n + 64);
    let mut frames: Vec<(usize, usize)> = Vec::with_capacity(n);
    for r in records.iter().skip(cfg.warmup) {
        let start = journal.len();
        let len = encode(r, &mut scratch);
        journal.extend_from_slice(&scratch[..len]);
        frames.push((start, len));
    }

    let mut decode_ns: Vec<u64> = Vec::with_capacity(n);
    let mut sink = 0u64;

    // Warmup decode (also warms any lazy statics before we start counting).
    for &(off, len) in frames.iter().take(cfg.warmup.min(frames.len())) {
        sink ^= decode(&journal[off..off + len]);
    }

    reset_allocated();
    let alloc_before = allocated_bytes();
    for &(off, len) in &frames {
        let t0 = Instant::now();
        let sum = decode(&journal[off..off + len]);
        let dt = t0.elapsed().as_nanos() as u64;
        sink ^= sum;
        decode_ns.push(dt);
    }
    let alloc_after = allocated_bytes();
    black_box(sink);

    let decode_alloc_per = (alloc_after - alloc_before) / n.max(1);

    emit_latency(experiment, "encode", &encode_ns);
    emit_latency(experiment, "decode", &decode_ns);
    result::emit(
        FOCUS,
        experiment,
        "encoded_bytes",
        record_len as u64,
        "bytes",
        1,
    );
    result::emit(
        FOCUS,
        experiment,
        "decode_alloc_bytes",
        decode_alloc_per as u64,
        "bytes",
        n,
    );
}

/// p50/p99 (integer ns) and mean (fractional ns) over `samples`. Sorts a copy
/// (percentiles need sorted input) and delegates to the shared `stats` module;
/// pure so it can be unit-tested without capturing stdout.
fn latency_stats(samples: &[u64]) -> (u64, u64, f64) {
    let mut sorted = samples.to_vec();
    sorted.sort_unstable();
    let p50 = stats::percentile(&sorted, 50.0);
    let p99 = stats::percentile(&sorted, 99.0);
    (p50, p99, stats::mean(samples))
}

/// Emit `<op>_p50`, `<op>_p99` (ns) and `<op>_mean` (ns) for one operation's
/// per-iteration latency samples.
fn emit_latency(experiment: &str, op: &str, samples: &[u64]) {
    let n = samples.len();
    let (p50, p99, mean) = latency_stats(samples);
    result::emit(FOCUS, experiment, &format!("{op}_p50"), p50, "ns", n);
    result::emit(FOCUS, experiment, &format!("{op}_p99"), p99, "ns", n);
    result::emit_float(FOCUS, experiment, &format!("{op}_mean"), mean, "ns", n);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn config_defaults_from_env() {
        // No SER_* vars set in the test environment → from_env yields the
        // field-heavy record defaults. (edition 2024 makes set_var unsafe, so
        // we test the default path rather than mutating the environment.)
        let cfg = SerialConfig::from_env().expect("defaults");
        assert_eq!(cfg.warmup, 1000);
        assert_eq!(cfg.iterations, 100_000);
        assert_eq!(cfg.entries, 4);
        assert_eq!(cfg.cmd_bytes, 12);
    }

    #[test]
    fn counting_allocator_tracks_a_vec() {
        reset_allocated();
        let before = allocated_bytes();
        let v: Vec<u8> = Vec::with_capacity(4096);
        assert!(allocated_bytes() >= before + 4096);
        drop(v);
    }

    #[test]
    fn latency_stats_sorts_before_percentiles() {
        // Unsorted input must yield the same result as the sorted slice through
        // `stats` — this is the only logic `latency_stats` adds over `stats`.
        let unsorted = vec![50u64, 10, 99, 1, 30, 70, 20, 90, 40, 60];
        let (p50, p99, mean) = latency_stats(&unsorted);

        let mut sorted = unsorted.clone();
        sorted.sort_unstable();
        assert_eq!(p50, stats::percentile(&sorted, 50.0));
        assert_eq!(p99, stats::percentile(&sorted, 99.0));
        assert_eq!(mean, stats::mean(&unsorted));
        assert!(p50 <= p99, "p50 {p50} must not exceed p99 {p99}");
    }
}

File: go/internal/bench/serial.go
package bench

import (
	"runtime"
	"sort"
	"time"
)

const serFocusArea = "serialization"

// serSink absorbs the decode checksums so the fold cannot be elided.
var serSink uint64

// SerialConfig configures the serialization journal benchmark, sourced from
// SER_* env vars (same names and defaults as the Rust cells). CmdBytes sizes
// the cmdText string on each entry (default 12); the record is field-heavy,
// dominated by the typed scalar fields rather than cmdText.
type SerialConfig struct {
	Warmup   int
	Iters    int
	Entries  int
	CmdBytes int
}

// LoadSerialConfig reads and validates the SER_* environment.
func LoadSerialConfig() (SerialConfig, error) {
	warmup, err := positiveEnv("SER_WARMUP", 1000)
	if err != nil {
		return SerialConfig{}, err
	}
	iters, err := positiveEnv("SER_ITERS", 100000)
	if err != nil {
		return SerialConfig{}, err
	}
	entries, err := positiveEnv("SER_ENTRIES", 4)
	if err != nil {
		return SerialConfig{}, err
	}
	cmdBytes, err := positiveEnv("SER_CMD_BYTES", 12)
	if err != nil {
		return SerialConfig{}, err
	}
	return SerialConfig{Warmup: warmup, Iters: iters, Entries: entries, CmdBytes: cmdBytes}, nil
}

// RunJournal drives the journal write/replay loop and emits the eight
// result-contract metrics. build(index) produces one codec-native record
// deterministically (pre-built, untimed — conversion from the logical model is
// not part of the measurement); encode(record, scratch) serializes into the
// reused scratch buffer and returns the encoded length; decode(bytes) decodes
// and fully materializes every field into a checksum, so owned-decode codecs
// and any future lazy codec pay for the same reads.
//
// Generic over the record type R so this package stays focus-neutral and never
// imports a focus area's model package.
func RunJournal[R any](experiment string, cfg SerialConfig, build func(uint64) R, encode func(R, []byte) int, decode func([]byte) uint64) {
	n := cfg.Iters

	// Pre-build all records (untimed); building is deterministic from index.
	records := make([]R, cfg.Warmup+n)
	for i := range records {
		records[i] = build(uint64(i))
	}

	scratch := make([]byte, 64*1024)
	encodeNs := make([]int64, 0, n)
	recordLen := 0

	// Warmup encode.
	for _, r := range records[:cfg.Warmup] {
		recordLen = encode(r, scratch)
	}
	// Timed encode.
	for _, r := range records[cfg.Warmup:] {
		t0 := time.Now()
		l := encode(r, scratch)
		dt := time.Since(t0).Nanoseconds()
		serSink ^= uint64(scratch[0])
		recordLen = l
		encodeNs = append(encodeNs, dt)
	}

	// Build the contiguous in-memory journal from the timed records.
	type frame struct{ off, len int }
	journal := make([]byte, 0, recordLen*n+64)
	frames := make([]frame, 0, n)
	for _, r := range records[cfg.Warmup:] {
		start := len(journal)
		l := encode(r, scratch)
		journal = append(journal, scratch[:l]...)
		frames = append(frames, frame{start, l})
	}

	decodeNs := make([]int64, 0, n)
	var sink uint64

	// Warmup decode.
	warm := cfg.Warmup
	if warm > len(frames) {
		warm = len(frames)
	}
	for _, f := range frames[:warm] {
		sink ^= decode(journal[f.off : f.off+f.len])
	}

	var before, after runtime.MemStats
	runtime.ReadMemStats(&before)
	for _, f := range frames {
		t0 := time.Now()
		sum := decode(journal[f.off : f.off+f.len])
		dt := time.Since(t0).Nanoseconds()
		sink ^= sum
		decodeNs = append(decodeNs, dt)
	}
	runtime.ReadMemStats(&after)
	serSink ^= sink

	decodeAllocPer := (after.TotalAlloc - before.TotalAlloc) / uint64(n)

	emitSerLatency(experiment, "encode", encodeNs)
	emitSerLatency(experiment, "decode", decodeNs)
	Emit(Result{FocusArea: serFocusArea, Experiment: experiment, Metric: "encoded_bytes",
		Value: float64(recordLen), Unit: "bytes", Samples: 1})
	Emit(Result{FocusArea: serFocusArea, Experiment: experiment, Metric: "decode_alloc_bytes",
		Value: float64(decodeAllocPer), Unit: "bytes", Samples: int64(n)})
}

// emitSerLatency sorts samples and emits {op}_p50/p99/mean (ns), mirroring
// EmitSmrLatency but for the serialization focus area.
func emitSerLatency(experiment, op string, samples []int64) {
	sort.Slice(samples, func(i, j int) bool { return samples[i] < samples[j] })
	n := int64(len(samples))
	Emit(Result{FocusArea: serFocusArea, Experiment: experiment, Metric: op + "_p50", Value: float64(Percentile(samples, 50)), Unit: "ns", Samples: n})
	Emit(Result{FocusArea: serFocusArea, Experiment: experiment, Metric: op + "_p99", Value: float64(Percentile(samples, 99)), Unit: "ns", Samples: n})
	Emit(Result{FocusArea: serFocusArea, Experiment: experiment, Metric: op + "_mean", Value: Mean(samples), Unit: "ns", Samples: n})
}
